App crash without secure lock screen

```
E  FATAL EXCEPTION: main
Process: at.asitplus.wallet.app.android, PID: 27092
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:603)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:932)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:593)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:932) 
Caused by: java.security.InvalidAlgorithmParameterException: java.lang.IllegalStateException: Secure lock screen must be enabled to create keys requiring user authentication
	at android.security.keystore2.AndroidKeyStoreKeyPairGeneratorSpi.initialize(AndroidKeyStoreKeyPairGeneratorSpi.java:368)
	at java.security.KeyPairGenerator$Delegate.initialize(KeyPairGenerator.java:720)
	at java.security.KeyPairGenerator.initialize(KeyPairGenerator.java:461)
	at at.asitplus.signum.supreme.os.AndroidKeyStoreProvider$createSigningKey$2.invokeSuspend(AndroidKeyStoreProvider.kt:220)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1156)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:651)
	at java.lang.Thread.run(Thread.java:1119)
Caused by: java.lang.IllegalStateException: Secure lock screen must be enabled to create keys requiring user authentication
	at android.security.keystore2.KeyStore2ParameterUtils.getRootSid(KeyStore2ParameterUtils.java:237)
	at android.security.keystore2.KeyStore2ParameterUtils.addSids(KeyStore2ParameterUtils.java:255)
	at android.security.keystore2.KeyStore2ParameterUtils.addUserAuthArgs(KeyStore2ParameterUtils.java:330)
	at android.security.keystore2.AndroidKeyStoreKeyPairGeneratorSpi.initialize(AndroidKeyStoreKeyPairGeneratorSpi.java:366)
	at java.security.KeyPairGenerator$Delegate.initialize(KeyPairGenerator.java:720) 
	at java.security.KeyPairGenerator.initialize(KeyPairGenerator.java:461) 
	at at.asitplus.signum.supreme.os.AndroidKeyStoreProvider$createSigningKey$2.invokeSuspend(AndroidKeyStoreProvider.kt:220) 
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100) 
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1156) 
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:651) 
	at java.lang.Thread.run(Thread.java:1119) 
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,8 @@
  * Update to VC-K 5.8.0, fixing optional attributes during presentation
  * Credentials: Add FallbackCredentialScheme for unknown schemes
  * Improve display of complex credentials in technical detail view
+ * Add SessionService to handle scoped Koin objects
+ * Add FallBackKeyMaterial to catch startup exceptions
 
 # Release 5.7.2 
  * Proximity: Use fixed IACA key and certificate for reader authentication

--- a/shared/src/androidMain/kotlin/at/asitplus/wallet/app/android/KeystoreProvider.android.kt
+++ b/shared/src/androidMain/kotlin/at/asitplus/wallet/app/android/KeystoreProvider.android.kt
@@ -1,7 +1,0 @@
-package at.asitplus.wallet.app.common
-
-import at.asitplus.signum.supreme.os.PlatformSigningProvider
-import at.asitplus.signum.supreme.os.SigningProvider
-
-
-actual fun getProvider(): SigningProvider = PlatformSigningProvider

--- a/shared/src/androidMain/kotlin/main.android.kt
+++ b/shared/src/androidMain/kotlin/main.android.kt
@@ -37,6 +37,7 @@ import at.asitplus.wallet.app.common.BuildContext
 import at.asitplus.wallet.app.common.KeystoreService
 import at.asitplus.wallet.app.common.PlatformAdapter
 import at.asitplus.wallet.app.common.WalletDependencyProvider
+import at.asitplus.wallet.app.common.WalletKeyMaterial
 import at.asitplus.wallet.app.common.dcapi.data.export.CredentialList
 import at.asitplus.wallet.app.common.dcapi.data.preview.ResponseJSON
 import at.asitplus.wallet.lib.data.vckJsonSerializer
@@ -96,7 +97,7 @@ fun MainView(
 
     App(
         WalletDependencyProvider(
-            keyMaterial = ks.let { runBlocking { AndroidKeyMaterial(it.getSigner()) } },
+            keystoreService = ks,
             dataStoreService = dataStoreService,
             platformAdapter = platformAdapter,
             buildContext = buildContext,
@@ -391,4 +392,8 @@ public class AndroidPlatformAdapter(
 actual fun getImageDecoder(image: ByteArray): ImageBitmap {
     val bitmap = BitmapFactory.decodeByteArray(image, 0, image.size)
     return bitmap.asImageBitmap()
+}
+
+actual fun getKeyMaterial(keystoreService: KeystoreService): WalletKeyMaterial {
+    return AndroidKeyMaterial(keystoreService.getSignerBlocking())
 }

--- a/shared/src/commonMain/kotlin/App.kt
+++ b/shared/src/commonMain/kotlin/App.kt
@@ -1,12 +1,11 @@
 import androidx.compose.material3.ColorScheme
 import androidx.compose.runtime.Composable
+import androidx.compose.runtime.collectAsState
 import androidx.compose.ui.graphics.ImageBitmap
 import androidx.lifecycle.Lifecycle
 import androidx.lifecycle.compose.LifecycleEventEffect
 import at.asitplus.catchingUnwrapped
-import at.asitplus.wallet.app.common.ErrorService
-import at.asitplus.wallet.app.common.WalletDependencyProvider
-import at.asitplus.wallet.app.common.WalletMain
+import at.asitplus.wallet.app.common.*
 import at.asitplus.wallet.app.common.dcapi.DCAPIInvocationData
 import at.asitplus.wallet.app.common.di.appModule
 import io.github.aakira.napier.Napier
@@ -37,8 +36,9 @@ fun App(walletDependencyProvider: WalletDependencyProvider) {
     KoinApplication({
         modules(appModule(walletDependencyProvider))
     }) {
+        val koinScope = koinInject<SessionService>().scope.collectAsState().value
         catchingUnwrapped {
-            val walletMain: WalletMain = koinInject()
+            val walletMain: WalletMain = koinInject(scope = koinScope)
 
             LifecycleEventEffect(Lifecycle.Event.ON_CREATE) {
                 Napier.d("Lifecycle.Event.ON_CREATE")
@@ -50,12 +50,12 @@ fun App(walletDependencyProvider: WalletDependencyProvider) {
             }
 
         }.onFailure {
-            val errorService: ErrorService = koinInject()
+            val errorService: ErrorService = koinInject(scope = koinScope)
             errorService.emit(it)
         }
 
         WalletTheme {
-            WalletNavigation()
+            WalletNavigation(koinScope = koinScope)
         }
     }
 }
@@ -66,3 +66,5 @@ expect fun getPlatformName(): String
 expect fun getColorScheme(): ColorScheme
 
 expect fun getImageDecoder(image: ByteArray): ImageBitmap
+
+expect fun getKeyMaterial(keystoreService: KeystoreService): WalletKeyMaterial

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
@@ -154,6 +154,10 @@ open class KeystoreService(
     fun getSignerBlocking() = runBlocking { getSigner() }
 }
 
+/**
+ * Fallback class if KeyMaterial initialization throws on startup.
+ * Allows to handle errors with the default navigation procedure.
+ */
 class FallBackKeyMaterial(
     val reason: Throwable,
     override val signatureAlgorithm: SignatureAlgorithm = SignatureAlgorithm.ECDSAwithSHA256,

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
@@ -1,15 +1,19 @@
 package at.asitplus.wallet.app.common
 
 import AppResetRequiredException
+import at.asitplus.KmmResult
+import at.asitplus.catchingUnwrapped
 import at.asitplus.io.MultiBase
 import at.asitplus.io.multibaseDecode
 import at.asitplus.io.multibaseEncode
-import at.asitplus.signum.indispensable.ECCurve
-import at.asitplus.signum.indispensable.equalsCryptographically
+import at.asitplus.signum.indispensable.*
 import at.asitplus.signum.indispensable.pki.X509Certificate
+import at.asitplus.signum.supreme.SignatureResult
 import at.asitplus.signum.supreme.dsl.PREFERRED
 import at.asitplus.signum.supreme.os.SigningProvider
+import at.asitplus.signum.supreme.sign.SignatureInput
 import at.asitplus.signum.supreme.sign.Signer
+import at.asitplus.wallet.lib.agent.EphemeralKeyWithoutCert
 import at.asitplus.wallet.lib.agent.KeyMaterial
 import at.asitplus.wallet.lib.agent.KeyWithSelfSignedCert
 import data.storage.DataStoreService
@@ -34,7 +38,7 @@ open class KeystoreService(
         Napier.d("getSigner")
         sMut.withLock {
             if (signer == null)
-                signer = initSigner()
+                signer = catchingUnwrapped { initSigner() }.getOrElse { FallBackKeyMaterial() }
         }
         return signer!!
     }
@@ -146,6 +150,30 @@ open class KeystoreService(
     //TMP for iOS
     @Throws(Throwable::class)
     fun getSignerBlocking() = runBlocking { getSigner() }
+}
+
+class FallBackKeyMaterial(
+    override val signatureAlgorithm: SignatureAlgorithm = SignatureAlgorithm.ECDSAwithSHA256,
+    override val publicKey: CryptoPublicKey = EphemeralKeyWithoutCert().publicKey,
+    override val identifier: String = ""
+): KeyMaterial {
+    @SecretExposure
+    override fun exportPrivateKey(): KmmResult<CryptoPrivateKey.WithPublicKey<*>> {
+        throw Throwable("Not intended for usage")
+    }
+
+    override suspend fun sign(data: SignatureInput): SignatureResult<*> {
+        throw Throwable("Not intended for usage")
+    }
+
+    override fun getUnderLyingSigner(): Signer {
+        throw Throwable("Not intended for usage")
+    }
+
+    override suspend fun getCertificate(): X509Certificate? {
+        throw Throwable("Not intended for usage")
+    }
+
 }
 
 expect fun getProvider(): SigningProvider

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
@@ -10,6 +10,7 @@ import at.asitplus.signum.indispensable.*
 import at.asitplus.signum.indispensable.pki.X509Certificate
 import at.asitplus.signum.supreme.SignatureResult
 import at.asitplus.signum.supreme.dsl.PREFERRED
+import at.asitplus.signum.supreme.os.PlatformSigningProvider
 import at.asitplus.signum.supreme.os.SigningProvider
 import at.asitplus.signum.supreme.sign.SignatureInput
 import at.asitplus.signum.supreme.sign.Signer
@@ -38,14 +39,15 @@ open class KeystoreService(
         Napier.d("getSigner")
         sMut.withLock {
             if (signer == null)
-                signer = catchingUnwrapped { initSigner() }.getOrElse { FallBackKeyMaterial() }
+                signer = catchingUnwrapped { initSigner() }
+                    .getOrElse { FallBackKeyMaterial(it) }
         }
         return signer!!
     }
 
     @Throws(Throwable::class)
     private suspend fun initSigner(): KeyWithSelfSignedCert {
-        getProvider().let { provider ->
+        PlatformSigningProvider.let { provider ->
             val forKey = provider.getSignerForKey(Configuration.KS_ALIAS).getOrElse {
                 provider.createSigningKey(alias = Configuration.KS_ALIAS) {
                     ec {
@@ -121,7 +123,7 @@ open class KeystoreService(
     companion object {
         @Throws(AppResetRequiredException::class)
         fun checkKeyMaterialValid() {
-            getProvider().let { provider ->
+            PlatformSigningProvider.let { provider ->
                 runBlocking {
                     provider.getSignerForKey(Configuration.KS_ALIAS_OLD).onSuccess {
                         provider.deleteSigningKey(Configuration.KS_ALIAS_OLD)
@@ -134,7 +136,7 @@ open class KeystoreService(
         }
 
         fun clearKeyMaterial() {
-            getProvider().let { provider ->
+            PlatformSigningProvider.let { provider ->
                 runBlocking {
                     provider.getSignerForKey(Configuration.KS_ALIAS_OLD).onSuccess {
                         provider.deleteSigningKey(Configuration.KS_ALIAS_OLD)
@@ -153,13 +155,14 @@ open class KeystoreService(
 }
 
 class FallBackKeyMaterial(
+    val reason: Throwable,
     override val signatureAlgorithm: SignatureAlgorithm = SignatureAlgorithm.ECDSAwithSHA256,
     override val publicKey: CryptoPublicKey = EphemeralKeyWithoutCert().publicKey,
     override val identifier: String = ""
 ): KeyMaterial {
     @SecretExposure
     override fun exportPrivateKey(): KmmResult<CryptoPrivateKey.WithPublicKey<*>> {
-        throw Throwable("Not intended for usage")
+        throw NotImplementedError()
     }
 
     override suspend fun sign(data: SignatureInput): SignatureResult<*> {
@@ -175,5 +178,3 @@ class FallBackKeyMaterial(
     }
 
 }
-
-expect fun getProvider(): SigningProvider

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.kt
@@ -163,15 +163,15 @@ class FallBackKeyMaterial(
     }
 
     override suspend fun sign(data: SignatureInput): SignatureResult<*> {
-        throw Throwable("Not intended for usage")
+        throw NotImplementedError()
     }
 
     override fun getUnderLyingSigner(): Signer {
-        throw Throwable("Not intended for usage")
+        throw NotImplementedError()
     }
 
     override suspend fun getCertificate(): X509Certificate? {
-        throw Throwable("Not intended for usage")
+        throw NotImplementedError()
     }
 
 }

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/SessionService.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/SessionService.kt
@@ -8,6 +8,10 @@ import kotlin.uuid.ExperimentalUuidApi
 import kotlin.uuid.Uuid
 
 val SESSION_NAME = "WALLET_SESSION"
+/**
+ * Manages creation and deletion of Koin scopes.
+ * Allows to reinitialize singleton dependencies e.g. on App reset
+ */
 class SessionService(): KoinComponent {
     private val koin = getKoin()
 

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/SessionService.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/SessionService.kt
@@ -1,0 +1,29 @@
+package at.asitplus.wallet.app.common
+
+import kotlinx.coroutines.flow.MutableStateFlow
+import org.koin.core.component.KoinComponent
+import org.koin.core.qualifier.named
+import org.koin.core.scope.Scope
+import kotlin.uuid.ExperimentalUuidApi
+import kotlin.uuid.Uuid
+
+val SESSION_NAME = "WALLET_SESSION"
+class SessionService(): KoinComponent {
+    private val koin = getKoin()
+
+    private val scopeId = MutableStateFlow(generateUuid())
+    val scope = MutableStateFlow(initScope())
+
+
+    @OptIn(ExperimentalUuidApi::class)
+    fun generateUuid() = Uuid.random().toString()
+    private fun initScope(): Scope {
+        scopeId.value = generateUuid()
+        return this.koin.createScope(scopeId.value, named(SESSION_NAME))
+    }
+
+    fun newScope() {
+        koin.deleteScope(scopeId.value)
+        scope.value = initScope()
+    }
+}

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/SessionService.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/SessionService.kt
@@ -13,21 +13,19 @@ val SESSION_NAME = "WALLET_SESSION"
  * Allows to reinitialize singleton dependencies e.g. on App reset
  */
 class SessionService(): KoinComponent {
-    private val koin = getKoin()
-
-    private val scopeId = MutableStateFlow(generateUuid())
+    private var scopeId = generateUuid()
     val scope = MutableStateFlow(initScope())
 
-
     @OptIn(ExperimentalUuidApi::class)
-    fun generateUuid() = Uuid.random().toString()
+    private fun generateUuid() = Uuid.random().toString()
+
     private fun initScope(): Scope {
-        scopeId.value = generateUuid()
-        return this.koin.createScope(scopeId.value, named(SESSION_NAME))
+        scopeId = generateUuid()
+        return getKoin().createScope(scopeId, named(SESSION_NAME))
     }
 
     fun newScope() {
-        koin.deleteScope(scopeId.value)
+        getKoin().deleteScope(scopeId)
         scope.value = initScope()
     }
 }

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/WalletDependencyProvider.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/WalletDependencyProvider.kt
@@ -8,7 +8,7 @@ import io.github.aakira.napier.Napier
 import org.multipaz.prompt.PromptModel
 
 data class WalletDependencyProvider(
-    val keyMaterial: WalletKeyMaterial,
+    val keystoreService: KeystoreService,
     val dataStoreService: DataStoreService,
     val platformAdapter: PlatformAdapter,
     var subjectCredentialStore: PersistentSubjectCredentialStore =

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/WalletKeyMaterial.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/WalletKeyMaterial.kt
@@ -19,12 +19,13 @@ open class WalletKeyMaterial(val keyMaterial: KeyMaterial) :
         data: ByteArray
     ): SignatureResult<CryptoSignature.RawByteEncodable> = run {
         when (val signer = keyMaterial.getUnderLyingSigner()) {
-            is PlatformSigningProviderSigner<*, *> -> signer.sign(data) {
-                unlockPrompt {
-                    promptText?.let { message = it }
-                    promptCancelText?.let { cancelText = it }
+            is PlatformSigningProviderSigner<*, *> ->
+                signer.sign(data) {
+                    unlockPrompt {
+                        promptText?.let { message = it }
+                        promptCancelText?.let { cancelText = it }
+                    }
                 }
-            }
 
             else -> signer.sign(data)
         }.also {

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/WalletMain.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/WalletMain.kt
@@ -1,6 +1,5 @@
 package at.asitplus.wallet.app.common
 
-import androidx.compose.ui.graphics.ImageBitmap
 import at.asitplus.KmmResult
 import at.asitplus.catchingUnwrapped
 import at.asitplus.dcapi.request.DCAPIRequest
@@ -20,17 +19,12 @@ import data.storage.DataStoreService
 import data.storage.WalletSubjectCredentialStore
 import getImageDecoder
 import io.github.aakira.napier.Napier
-import io.ktor.client.call.body
-import io.ktor.client.request.get
-import kotlinx.coroutines.CoroutineExceptionHandler
-import kotlinx.coroutines.CoroutineName
-import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.IO
+import io.ktor.client.call.*
+import io.ktor.client.request.*
+import kotlinx.coroutines.*
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.launchIn
 import kotlinx.coroutines.flow.onEach
-import kotlinx.coroutines.launch
 import kotlinx.serialization.json.JsonObject
 import kotlinx.serialization.json.jsonObject
 import kotlinx.serialization.json.jsonPrimitive
@@ -45,9 +39,9 @@ class WalletMain(
     val keyMaterial: WalletKeyMaterial,
     val dataStoreService: DataStoreService,
     val platformAdapter: PlatformAdapter,
-    var subjectCredentialStore: WalletSubjectCredentialStore,
+    val subjectCredentialStore: WalletSubjectCredentialStore,
     val buildContext: BuildContext,
-    val promptModel: PromptModel,
+    promptModel: PromptModel,
     val credentialValidator: Validator,
     val holderAgent: HolderAgent,
     val provisioningService: ProvisioningService,
@@ -58,6 +52,7 @@ class WalletMain(
     val errorService: ErrorService,
     val snackbarService: SnackbarService,
     val settingsRepository: SettingsRepository,
+    val sessionService: SessionService,
 ) {
     val appReady = MutableStateFlow<Boolean?>(null)
 
@@ -87,6 +82,7 @@ class WalletMain(
         KeystoreService.clearKeyMaterial()
 
         settingsRepository.reset()
+        sessionService.newScope()
     }
 
     fun getLog(): List<String> {
@@ -166,8 +162,8 @@ class WalletMain(
 }
 
 fun PlatformAdapter.decodeImage(image: ByteArray) = catchingUnwrapped {
-        getImageDecoder((image))
-    }
+    getImageDecoder((image))
+}
 
 
 /**

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/di/AppModule.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/di/AppModule.kt
@@ -1,18 +1,25 @@
 package at.asitplus.wallet.app.common.di
 
 import at.asitplus.wallet.app.common.ErrorService
+import at.asitplus.wallet.app.common.SESSION_NAME
+import at.asitplus.wallet.app.common.SessionService
 import at.asitplus.wallet.app.common.WalletDependencyProvider
 import at.asitplus.wallet.app.common.WalletMain
 import at.asitplus.wallet.app.common.data.di.dataModule
 import at.asitplus.wallet.app.common.domain.di.domainModule
+import org.koin.core.module.dsl.scopedOf
 import org.koin.core.module.dsl.singleOf
+import org.koin.core.qualifier.named
 import org.koin.dsl.module
 import ui.navigation.IntentService
 
 fun appModule(appDependencyProvider: WalletDependencyProvider) = module {
-    singleOf(::WalletMain)
+    scope(named(SESSION_NAME)) {
+        scopedOf(::WalletMain)
+    }
     singleOf(::ErrorService)
     singleOf(::IntentService)
+    singleOf(::SessionService)
     includes(dataModule())
     includes(uiModule())
     includes(domainModule(appDependencyProvider))

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/communications/di/CommunicationsModule.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/communications/di/CommunicationsModule.kt
@@ -3,15 +3,20 @@ package at.asitplus.wallet.app.common.domain.communications.di
 import at.asitplus.wallet.app.common.HttpService
 import at.asitplus.wallet.app.common.PresentationService
 import at.asitplus.wallet.app.common.ProvisioningService
+import at.asitplus.wallet.app.common.SESSION_NAME
 import at.asitplus.wallet.app.common.SigningService
 import at.asitplus.wallet.app.common.dcapi.DCAPIExportService
+import org.koin.core.module.dsl.scopedOf
 import org.koin.core.module.dsl.singleOf
+import org.koin.core.qualifier.named
 import org.koin.dsl.module
 
 fun communicationsModule() = module {
     singleOf(::HttpService)
-    singleOf(::ProvisioningService)
-    singleOf(::PresentationService)
+    scope(named(SESSION_NAME)) {
+        scopedOf(::ProvisioningService)
+        scopedOf(::PresentationService)
+    }
     singleOf(::SigningService)
     singleOf(::DCAPIExportService)
 }

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/platform/di/PlatformModule.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/platform/di/PlatformModule.kt
@@ -2,6 +2,7 @@ package at.asitplus.wallet.app.common.domain.platform.di
 
 import at.asitplus.wallet.app.common.BuildContext
 import at.asitplus.wallet.app.common.PlatformAdapter
+import at.asitplus.wallet.app.common.SESSION_NAME
 import at.asitplus.wallet.app.common.WalletDependencyProvider
 import at.asitplus.wallet.app.common.WalletKeyMaterial
 import at.asitplus.wallet.app.common.decodeImage
@@ -13,17 +14,21 @@ import data.storage.DataStoreService
 import data.storage.HotWalletSubjectCredentialStore
 import data.storage.PersistentSubjectCredentialStore
 import data.storage.WalletSubjectCredentialStore
+import getKeyMaterial
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.IO
+import org.koin.core.qualifier.named
 import org.koin.dsl.binds
 import org.koin.dsl.module
 import org.multipaz.prompt.PromptModel
 
 fun platformModule(appDependencyProvider: WalletDependencyProvider) = module {
-    single<WalletKeyMaterial> {
-        appDependencyProvider.keyMaterial
-    } binds arrayOf(KeyMaterial::class)
+    scope(named(SESSION_NAME)) {
+        scoped<WalletKeyMaterial> {
+            getKeyMaterial(appDependencyProvider.keystoreService)
+        } binds arrayOf(KeyMaterial::class)
+    }
 
     factory<DataStoreService> {
         appDependencyProvider.dataStoreService

--- a/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/vck/di/VckModule.kt
+++ b/shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/vck/di/VckModule.kt
@@ -1,5 +1,6 @@
 package at.asitplus.wallet.app.common.domain.vck.di
 
+import at.asitplus.wallet.app.common.SESSION_NAME
 import at.asitplus.wallet.app.common.domain.vck.tokenStatusList.StatusListTokenResolver
 import at.asitplus.wallet.app.common.domain.vck.JsonWebKeySetResolver
 import at.asitplus.wallet.app.common.domain.vck.PublicKeyResolver
@@ -10,6 +11,7 @@ import at.asitplus.wallet.lib.agent.SubjectCredentialStore
 import at.asitplus.wallet.lib.agent.Validator
 import at.asitplus.wallet.lib.jws.VerifyJwsObject
 import org.koin.core.module.dsl.singleOf
+import org.koin.core.qualifier.named
 import org.koin.dsl.module
 
 fun vckModule() = module {
@@ -23,15 +25,17 @@ fun vckModule() = module {
             verifyJwsObject = VerifyJwsObject(publicKeyLookup = publicKeyResolver::invoke)
         )
     }
-    single<HolderAgent> {
-        val keyMaterial: KeyMaterial by inject()
-        val subjectCredentialStore: SubjectCredentialStore by inject()
-        val validator: Validator by inject()
-        HolderAgent(
-            keyMaterial = keyMaterial,
-            subjectCredentialStore = subjectCredentialStore,
-            validator = validator,
-        )
+    scope(named(SESSION_NAME)) {
+        scoped<HolderAgent> {
+            val keyMaterial: KeyMaterial by inject()
+            val subjectCredentialStore: SubjectCredentialStore by inject()
+            val validator: Validator by inject()
+            HolderAgent(
+                keyMaterial = keyMaterial,
+                subjectCredentialStore = subjectCredentialStore,
+                validator = validator,
+            )
+        }
     }
     includes(tokenStatusListModule())
 }

--- a/shared/src/commonMain/kotlin/ui/navigation/WalletNavigation.kt
+++ b/shared/src/commonMain/kotlin/ui/navigation/WalletNavigation.kt
@@ -12,11 +12,7 @@ import androidx.compose.material3.Scaffold
 import androidx.compose.material3.SnackbarHost
 import androidx.compose.material3.SnackbarHostState
 import androidx.compose.material3.SnackbarResult
-import androidx.compose.runtime.Composable
-import androidx.compose.runtime.LaunchedEffect
-import androidx.compose.runtime.collectAsState
-import androidx.compose.runtime.getValue
-import androidx.compose.runtime.remember
+import androidx.compose.runtime.*
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.platform.testTag
 import androidx.navigation.NavHostController
@@ -47,85 +43,18 @@ import kotlinx.coroutines.launch
 import kotlinx.coroutines.runBlocking
 import org.jetbrains.compose.resources.getString
 import org.koin.compose.koinInject
+import org.koin.core.scope.Scope
 import ui.composables.BottomBar
 import ui.composables.NavigationData
-import ui.navigation.routes.AddCredentialPreAuthnRoute
-import ui.navigation.routes.AddCredentialRoute
-import ui.navigation.routes.AuthenticationSuccessRoute
-import ui.navigation.routes.AuthenticationViewRoute
-import ui.navigation.routes.AuthorizationIntentRoute
-import ui.navigation.routes.CredentialDetailsRoute
-import ui.navigation.routes.DCAPIAuthenticationConsentRoute
-import ui.navigation.routes.DCAPIAuthorizationIntentRoute
-import ui.navigation.routes.ErrorIntentRoute
-import ui.navigation.routes.ErrorRoute
-import ui.navigation.routes.HomeScreenRoute
-import ui.navigation.routes.LoadCredentialRoute
-import ui.navigation.routes.LoadingRoute
-import ui.navigation.routes.LocalPresentationAuthenticationConsentRoute
-import ui.navigation.routes.LogRoute
-import ui.navigation.routes.OnboardingInformationRoute
-import ui.navigation.routes.OnboardingStartRoute
-import ui.navigation.routes.OnboardingWrapperTestTags
-import ui.navigation.routes.PresentDataRoute
-import ui.navigation.routes.PresentationIntentRoute
-import ui.navigation.routes.ProvisioningIntentRoute
-import ui.navigation.routes.QrCodeScannerRoute
-import ui.navigation.routes.Route
-import ui.navigation.routes.SettingsRoute
-import ui.navigation.routes.ShowQrCodeRoute
-import ui.navigation.routes.SigningCredentialIntentRoute
-import ui.navigation.routes.SigningIntentRoute
-import ui.navigation.routes.SigningPreloadIntentRoute
-import ui.navigation.routes.SigningQtspSelectionRoute
-import ui.navigation.routes.SigningServiceIntentRoute
-import ui.navigation.routes.VerifyDataRoute
-import ui.viewmodels.CredentialDetailsViewModel
-import ui.viewmodels.ErrorViewModel
-import ui.viewmodels.LoadCredentialViewModel
-import ui.viewmodels.LogViewModel
-import ui.viewmodels.QrCodeScannerMode
-import ui.viewmodels.QrCodeScannerViewModel
-import ui.viewmodels.SigningQtspSelectionViewModel
-import ui.viewmodels.authentication.AuthenticationViewModel
-import ui.viewmodels.authentication.DefaultAuthenticationViewModel
-import ui.viewmodels.authentication.NewDCAPIAuthenticationViewModel
-import ui.viewmodels.authentication.PresentationViewModel
-import ui.viewmodels.authentication.PreviewDCAPIAuthenticationViewModel
-import ui.viewmodels.intents.AuthorizationIntentViewModel
-import ui.viewmodels.intents.DCAPIAuthorizationIntentViewModel
-import ui.viewmodels.intents.ErrorIntentViewModel
-import ui.viewmodels.intents.PresentationIntentViewModel
-import ui.viewmodels.intents.ProvisioningIntentViewModel
-import ui.viewmodels.intents.SigningCredentialIntentViewModel
-import ui.viewmodels.intents.SigningIntentViewModel
-import ui.viewmodels.intents.SigningPreloadIntentViewModel
-import ui.viewmodels.intents.SigningServiceIntentViewModel
+import ui.navigation.routes.*
+import ui.viewmodels.*
+import ui.viewmodels.authentication.*
+import ui.viewmodels.intents.*
 import ui.viewmodels.iso.VerifierViewModel
-import ui.views.CredentialDetailsView
-import ui.views.CredentialsView
-import ui.views.ErrorView
-import ui.views.LoadCredentialView
-import ui.views.LoadingView
-import ui.views.LogView
-import ui.views.OnboardingInformationView
-import ui.views.OnboardingStartView
-import ui.views.PresentDataView
-import ui.views.QrCodeScannerView
-import ui.views.SelectIssuingServerView
-import ui.views.SettingsView
-import ui.views.SigningQtspSelectionView
+import ui.views.*
 import ui.views.authentication.AuthenticationSuccessView
 import ui.views.authentication.AuthenticationView
-import ui.views.intents.AuthorizationIntentView
-import ui.views.intents.DCAPIAuthorizationIntentView
-import ui.views.intents.ErrorIntentView
-import ui.views.intents.PresentationIntentView
-import ui.views.intents.ProvisioningIntentView
-import ui.views.intents.SigningCredentialIntentView
-import ui.views.intents.SigningIntentView
-import ui.views.intents.SigningPreloadIntentView
-import ui.views.intents.SigningServiceIntentView
+import ui.views.intents.*
 import ui.views.iso.ShowQrCodeView
 import ui.views.iso.verifier.VerifierView
 import ui.views.presentation.PresentationView
@@ -136,11 +65,12 @@ internal object NavigatorTestTags {
 
 @Composable
 fun WalletNavigation(
+    koinScope: Scope,
     settingsRepository: SettingsRepository = koinInject(),
     intentService: IntentService = koinInject(),
     snackbarService: SnackbarService = koinInject(),
-    errorService: ErrorService = koinInject(),
-    walletMain: WalletMain = koinInject(),
+    errorService: ErrorService = koinInject(scope = koinScope),
+    walletMain: WalletMain = koinInject(scope = koinScope),
     urlOpener: UrlOpener = koinInject(),
 ) {
     val navController: NavHostController = rememberNavController()
@@ -196,6 +126,8 @@ fun WalletNavigation(
                 popBackStack(HomeScreenRoute)
                 errorService.emit(e)
             },
+            walletMain,
+            koinScope = koinScope
         )
     }
 
@@ -247,9 +179,10 @@ private fun WalletNavHost(
     popBackStack: (Route) -> Unit,
     onClickLogo: () -> Unit,
     onError: (Throwable) -> Unit,
-    walletMain: WalletMain = koinInject(),
+    walletMain: WalletMain,
     intentService: IntentService = koinInject(),
     settingsRepository: SettingsRepository = koinInject(),
+    koinScope: Scope,
 ) {
     val currentHost by settingsRepository.host.collectAsState("")
     NavHost(
@@ -295,7 +228,8 @@ private fun WalletNavHost(
                         navigate = { route -> navigate(route) },
                         selected = NavigationData.HOME_SCREEN
                     )
-                }
+                },
+                koinScope = koinScope
             )
             LaunchedEffect(null) {
                 walletMain.scope.launch {
@@ -335,6 +269,7 @@ private fun WalletNavHost(
                     Globals.presentationStateModel.value = it
                     navigate(LocalPresentationAuthenticationConsentRoute("QR"))
                 },
+                koinScope = koinScope
             )
         }
 
@@ -511,6 +446,7 @@ private fun WalletNavHost(
                 navigateUp = navigateBack,
                 onClickLogo = onClickLogo,
                 onClickSettings = { navigate(SettingsRoute) },
+                koinScope = koinScope
             )
         }
 
@@ -522,6 +458,7 @@ private fun WalletNavHost(
                 onNavigateToLoadCredentialRoute = { host ->
                     navigate(LoadCredentialRoute(host))
                 },
+                koinScope = koinScope
             )
         }
 
@@ -624,6 +561,7 @@ private fun WalletNavHost(
                 onClickFAQs = null,
                 onClickDataProtectionPolicy = null,
                 onClickLicenses = null,
+                koinScope = koinScope
             )
         }
 

--- a/shared/src/commonMain/kotlin/ui/views/CredentialsView.kt
+++ b/shared/src/commonMain/kotlin/ui/views/CredentialsView.kt
@@ -28,6 +28,7 @@ import at.asitplus.valera.resources.content_description_navigate_to_settings
 import at.asitplus.valera.resources.heading_label_my_data_screen
 import org.jetbrains.compose.resources.stringResource
 import org.koin.compose.viewmodel.koinViewModel
+import org.koin.core.scope.Scope
 import ui.composables.CustomFloatingActionMenu
 import ui.composables.FloatingActionButtonHeightSpacer
 import ui.composables.Logo
@@ -45,7 +46,8 @@ fun CredentialsView(
     navigateToCredentialDetailsPage: (Long) -> Unit,
     onClickLogo: () -> Unit,
     onClickSettings: () -> Unit,
-    vm: CredentialsViewModel = koinViewModel(),
+    koinScope: Scope,
+    vm: CredentialsViewModel = koinViewModel(scope = koinScope),
     bottomBar: @Composable () -> Unit
 ) {
     val credentialsStatus by vm.storeContainer.collectAsState()

--- a/shared/src/commonMain/kotlin/ui/views/SelectIssuingServerView.kt
+++ b/shared/src/commonMain/kotlin/ui/views/SelectIssuingServerView.kt
@@ -32,6 +32,7 @@ import at.asitplus.valera.resources.heading_label_add_credential_screen
 import kotlinx.coroutines.runBlocking
 import org.jetbrains.compose.resources.stringResource
 import org.koin.compose.viewmodel.koinViewModel
+import org.koin.core.scope.Scope
 import ui.composables.Logo
 import ui.composables.buttons.ContinueButton
 import ui.composables.buttons.NavigateUpButton
@@ -45,7 +46,8 @@ fun SelectIssuingServerView(
     navigateUp: () -> Unit,
     onClickLogo: () -> Unit,
     onClickSettings: () -> Unit,
-    vm: AddCredentialViewModel = koinViewModel(),
+    koinScope: Scope,
+    vm: AddCredentialViewModel = koinViewModel(scope = koinScope),
 ) {
     val host by vm.hostString.collectAsState()
     var hostInput by rememberSaveable(host, stateSaver = TextFieldValue.Saver) {

--- a/shared/src/commonMain/kotlin/ui/views/SettingsView.kt
+++ b/shared/src/commonMain/kotlin/ui/views/SettingsView.kt
@@ -74,6 +74,7 @@ import org.jetbrains.compose.resources.getString
 import org.jetbrains.compose.resources.painterResource
 import org.jetbrains.compose.resources.stringResource
 import org.koin.compose.viewmodel.koinViewModel
+import org.koin.core.scope.Scope
 import ui.composables.CircularProgressIndicatorOverlay
 import ui.composables.DelayedComposable
 import ui.composables.Logo
@@ -94,7 +95,8 @@ fun SettingsView(
     onClickFAQs: (() -> Unit)?,
     onClickDataProtectionPolicy: (() -> Unit)?,
     onClickLicenses: (() -> Unit)?,
-    settingsViewModel: SettingsViewModel = koinViewModel(),
+    koinScope: Scope,
+    settingsViewModel: SettingsViewModel = koinViewModel(scope = koinScope),
 ) {
     var isLoading by rememberSaveable {
         mutableStateOf(false)

--- a/shared/src/commonMain/kotlin/ui/views/authentication/AuthenticationSuccessView.kt
+++ b/shared/src/commonMain/kotlin/ui/views/authentication/AuthenticationSuccessView.kt
@@ -31,6 +31,7 @@ import at.asitplus.valera.resources.heading_label_authentication_success
 import at.asitplus.valera.resources.info_text_authentication_success
 import org.jetbrains.compose.resources.stringResource
 import org.koin.compose.viewmodel.koinViewModel
+import org.koin.core.scope.Scope
 import ui.composables.Logo
 import ui.composables.buttons.ConcludeButton
 import ui.composables.buttons.NavigateUpButton
@@ -43,7 +44,8 @@ fun AuthenticationSuccessView(
     navigateUp: () -> Unit,
     onClickLogo: () -> Unit,
     onClickSettings: () -> Unit,
-    vm: AuthenticationSuccessViewModel = koinViewModel(),
+    koinScope: Scope,
+    vm: AuthenticationSuccessViewModel = koinViewModel(scope = koinScope),
 ) {
     Scaffold(
         topBar = {

--- a/shared/src/commonMain/kotlin/ui/views/iso/ShowQrCodeView.kt
+++ b/shared/src/commonMain/kotlin/ui/views/iso/ShowQrCodeView.kt
@@ -46,6 +46,7 @@ import kotlinx.coroutines.launch
 import kotlinx.io.bytestring.ByteString
 import org.jetbrains.compose.resources.stringResource
 import org.koin.compose.viewmodel.koinViewModel
+import org.koin.core.scope.Scope
 import org.multipaz.compose.permissions.rememberBluetoothPermissionState
 import org.multipaz.util.toBase64Url
 import ui.composables.Logo
@@ -64,7 +65,8 @@ fun ShowQrCodeView(
     onNavigateToPresentmentScreen: (PresentationStateModel) -> Unit,
     onClickLogo: () -> Unit,
     onClickSettings: () -> Unit,
-    vm: ShowQrCodeViewModel = koinViewModel(),
+    koinScope: Scope,
+    vm: ShowQrCodeViewModel = koinViewModel(scope = koinScope),
 ) {
     val blePermissionState = rememberBluetoothPermissionState()
     val showQrCode = remember { mutableStateOf<ByteString?>(null) }

--- a/shared/src/commonTest/kotlin/at/asitplus/wallet/app/InstrumentedTestsSuite.kt
+++ b/shared/src/commonTest/kotlin/at/asitplus/wallet/app/InstrumentedTestsSuite.kt
@@ -6,6 +6,7 @@ import androidx.compose.runtime.Composable
 import androidx.compose.runtime.CompositionLocalProvider
 import androidx.compose.runtime.collectAsState
 import androidx.compose.runtime.getValue
+import androidx.compose.runtime.key
 import androidx.compose.ui.test.ExperimentalTestApi
 import androidx.compose.ui.test.assertHeightIsAtLeast
 import androidx.compose.ui.test.assertIsDisplayed
@@ -37,6 +38,7 @@ import at.asitplus.wallet.app.common.BuildContext
 import at.asitplus.wallet.app.common.BuildType
 import at.asitplus.wallet.app.common.KeystoreService
 import at.asitplus.wallet.app.common.PlatformAdapter
+import at.asitplus.wallet.app.common.SessionService
 import at.asitplus.wallet.app.common.WalletDependencyProvider
 import at.asitplus.wallet.app.common.WalletKeyMaterial
 import at.asitplus.wallet.eupid.EuPidScheme
@@ -220,17 +222,18 @@ class InstrumentedTestsSuite : FunSpec({
     context("End to End Tests") {
         test("End to End Test 1: Should complete the process") {
             runComposeUiTest {
-                lateinit var walletMain: WalletDependencyProvider
+                lateinit var walletDependencyProvider: WalletDependencyProvider
                 setContent {
                     CompositionLocalProvider(
                         LocalLifecycleOwner provides TestLifecycleOwner()
                     ) {
                         val platformAdapter = getPlatformAdapter()
-                        walletMain = createWalletDependencyProvider(platformAdapter)
-                        App(walletMain)
+                        walletDependencyProvider = createWalletDependencyProvider(platformAdapter)
+                        App(walletDependencyProvider)
                     }
+                    val sessionService: SessionService = koinInject()
+                    val holderAgent: HolderAgent = koinInject(scope = sessionService.scope.value)
 
-                    val holderAgent: HolderAgent = koinInject()
 
                     val keyMaterial = EphemeralKeyWithoutCert()
                     val issuer = IssuerAgent(
@@ -245,7 +248,7 @@ class InstrumentedTestsSuite : FunSpec({
                                     getAttributes(),
                                     Clock.System.now().plus(60.minutes),
                                     EuPidScheme,
-                                    walletMain.keyMaterial.keyMaterial.publicKey,
+                                    holderAgent.keyMaterial.publicKey,
                                     OidcUserInfoExtended(userInfo = OidcUserInfo(subject = ""))
                                 )
                             ).getOrThrow().toStoreCredentialInput()
@@ -367,7 +370,7 @@ private fun createWalletDependencyProvider(platformAdapter: PlatformAdapter): Wa
         override suspend fun getSigner(): KeyMaterial = EphemeralKeyWithSelfSignedCert()
     }
     return WalletDependencyProvider(
-        keyMaterial = ks.let { runBlocking { WalletKeyMaterial(it.getSigner()) } },
+        keystoreService = ks,
         dataStoreService = dummyDataStoreService,
         platformAdapter = platformAdapter,
         buildContext = BuildContext(

--- a/shared/src/iosMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.ios.kt
+++ b/shared/src/iosMain/kotlin/at/asitplus/wallet/app/common/KeystoreProvider.ios.kt
@@ -1,7 +1,0 @@
-package at.asitplus.wallet.app.common
-
-
-import at.asitplus.signum.supreme.os.IosKeychainProvider
-import at.asitplus.signum.supreme.os.SigningProvider
-
-actual fun getProvider(): SigningProvider = IosKeychainProvider

--- a/shared/src/iosMain/kotlin/main.ios.kt
+++ b/shared/src/iosMain/kotlin/main.ios.kt
@@ -47,7 +47,7 @@ fun MainViewController(
         PromptDialogs(promptModel)
         App(
             WalletDependencyProvider(
-                WalletKeyMaterial(keyMaterial = keystoreService.getSignerBlocking()),
+                keystoreService,
                 dataStoreService,
                 platformAdapter,
                 buildContext =  buildContext,
@@ -55,4 +55,8 @@ fun MainViewController(
             )
         )
     }
+}
+
+actual fun getKeyMaterial(keystoreService: KeystoreService): WalletKeyMaterial {
+    return WalletKeyMaterial(keystoreService.getSignerBlocking())
 }

